Panic in directory_resolver indexPath due to null info parameter

**What happened**:
Processing a particular tar file I received random panics in directory_resolver indexPath when the info parameter pasted to indexPath is null.

Note in syft/source/directory_resolver.go:138 a check is not made to make sure info != null   resulting in crash, but this is checked later in the function line 154.

panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x2 addr=0x18 pc=0x1009cd328]

goroutine 13 [running]:
github.com/anchore/syft/syft/source.(*directoryResolver).indexPath(0x140000e63f0, {0x140001a61c0, 0xf}, {0x0, 0x0}, {0x102985680, 0x1400bf403c0})
	/Users/runner/work/syft/syft/syft/source/directory_resolver.go:138 +0xa8
github.com/anchore/syft/syft/source.(*directoryResolver).indexTree.func1({0x140001a61c0, 0xf}, {0x0, 0x0}, {0x102985680, 0x1400bf403c0})
	/Users/runner/work/syft/syft/syft/source/directory_resolver.go:120 +0x88
path/filepath.Walk({0x140001a61c0, 0xf}, 0x1400085d8f8)
	/Users/runner/hostedtoolcache/go/1.17.7/x64/src/path/filepath/path.go:503 +0x68
github.com/anchore/syft/syft/source.(*directoryResolver).indexTree(0x140000e63f0, {0x140001a61c0, 0xf}, 0x14000f85100)
	/Users/runner/work/syft/syft/syft/source/directory_resolver.go:116 +0x280
github.com/anchore/syft/syft/source.indexAllRoots({0x1400018e240, 0x59}, 0x1400085dc10)
	/Users/runner/work/syft/syft/syft/source/directory_resolver.go:533 +0x14c
github.com/anchore/syft/syft/source.newDirectoryResolver({0x140008b2120, 0x51}, {0x0, 0x0, 0x0})
	/Users/runner/work/syft/syft/syft/source/directory_resolver.go:88 +0x424
github.com/anchore/syft/syft/source.(*Source).FileResolver(0x14001808000, {0x101e37a86, 0x8})
	/Users/runner/work/syft/syft/syft/source/source.go:301 +0x3e4
github.com/anchore/syft/syft.CatalogPackages(0x14001808000, {{0x1, 0x0, {0x101e37a86, 0x8}}})
	/Users/runner/work/syft/syft/syft/lib.go:38 +0x38
github.com/anchore/syft/cmd.generateCatalogPackagesTask.func1(0x1400019c420, 0x14001808000)
	/Users/runner/work/syft/syft/cmd/tasks.go:48 +0x74
github.com/anchore/syft/cmd.runTask(0x102956ea8, 0x1400019c420, 0x14001808000, 0x1400018e180, 0x140001e8000)
	/Users/runner/work/syft/syft/cmd/tasks.go:222 +0x6c
created by github.com/anchore/syft/cmd.buildRelationships
	/Users/runner/work/syft/syft/cmd/packages.go:282 +0x70


**What you expected to happen**:
Should check for info != null before calling info.IsDir()

**How to reproduce it (as minimally and precisely as possible)**:
Seemed to be random files - tar was 2GB so pretty large. Perhaps related to doing this on Mac / Mac file system

**Anything else we need to know?**:

**Environment**:
- Output of `syft version`:
Application:     syft
Version:         0.40.1
BuildDate:       2022-03-04T15:37:59Z
GitCommit:       b2ab4671b915d50b523ae77b85e9554e5f8bcf95
GitDescription:  v0.40.1
Platform:        darwin/arm64
GoVersion:       go1.17.7
Compiler:        gc

- OS (e.g: `cat /etc/os-release` or similar):
MacOS 12.2.1 M1 (ARM)

--- a/syft/source/directory_resolver.go
+++ b/syft/source/directory_resolver.go
@@ -132,10 +132,15 @@ func (r *directoryResolver) indexTree(root string, stager *progress.Stage) ([]st
 }
 
 func (r *directoryResolver) indexPath(path string, info os.FileInfo, err error) (string, error) {
+	// link cycles could cause a revisit --we should not allow this
+	if r.hasBeenIndexed(path) {
+		return "", nil
+	}
+
 	// ignore any path which a filter function returns true
 	for _, filterFn := range r.pathFilterFns {
 		if filterFn != nil && filterFn(path, info) {
-			if info.IsDir() {
+			if info != nil && info.IsDir() {
 				return "", fs.SkipDir
 			}
 			return "", nil
@@ -143,11 +148,6 @@ func (r *directoryResolver) indexPath(path string, info os.FileInfo, err error) 
 	}
 
 	if r.isFileAccessErr(path, err) {
-		return "", nil
-	}
-
-	// link cycles could cause a revisit --we should not allow this
-	if r.hasBeenIndexed(path) {
 		return "", nil
 	}
 

--- a/syft/source/directory_resolver_test.go
+++ b/syft/source/directory_resolver_test.go
@@ -865,3 +865,24 @@ func Test_IncludeRootPathInIndex(t *testing.T) {
 	_, exists = resolver.metadata[ref.ID()]
 	require.True(t, exists)
 }
+
+func TestDirectoryResolver_indexPath(t *testing.T) {
+	// TODO: Ideally we can use an OS abstraction, which would obviate the need for real FS setup.
+	tempFile, err := os.CreateTemp("", "")
+	require.NoError(t, err)
+
+	resolver, err := newDirectoryResolver(tempFile.Name())
+	require.NoError(t, err)
+
+	t.Run("filtering path with nil os.FileInfo", func(t *testing.T) {
+		// We use one of these prefixes in order to trigger a pathFilterFn
+		filteredPath := unixSystemRuntimePrefixes[0]
+
+		var fileInfo os.FileInfo = nil
+
+		assert.NotPanics(t, func() {
+			_, err := resolver.indexPath(filteredPath, fileInfo, nil)
+			assert.NoError(t, err)
+		})
+	})
+}

